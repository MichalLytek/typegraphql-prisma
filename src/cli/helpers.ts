import { GeneratorOptions } from "@prisma/generator-helper";
import { DMMF } from "../generator/types";

export async function loadFileCallback(
  filePath: string | undefined,
):
  | undefined
  | ((
      actionName: DMMF.ModelAction,
      typeName: string,
      options: GeneratorOptions,
    ) => string) {
  if (!filePath) {
    return;
  }

  const callback = await import(filePath);
  if (typeof callback !== "function") {
    throw new Error(`Invalid callback, callback must be a function`);
  }

  return callback;
}

export function parseStringBoolean(stringBoolean: string | undefined) {
  return stringBoolean ? stringBoolean === "true" : undefined;
}

export function parseStringArray<TAllowedValue extends string>(
  stringArray: string | undefined,
  optionPropertyName: string,
  allowedValues?: TAllowedValue[],
): TAllowedValue[] | undefined {
  if (!stringArray) {
    return undefined;
  }
  const parsedArray = stringArray.split(",").map(it => it.trim());
  if (allowedValues) {
    for (const option of parsedArray) {
      if (!allowedValues.includes(option as any)) {
        throw new Error(
          `Invalid "${optionPropertyName}" option value "${option}" provided for TypeGraphQL generator.`,
        );
      }
    }
  }
  return parsedArray as TAllowedValue[];
}
