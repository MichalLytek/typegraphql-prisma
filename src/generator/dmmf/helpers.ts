export const modelAttributeRegex = /(@@TypeGraphQL\.)+([A-z])+(\()+(.+)+(\))+/;
export const fieldAttributeRegex = /(@TypeGraphQL\.)+([A-z])+(\()+(.+)+(\))+/;
export const attributeNameRegex = /(?:\.)+([A-Za-z])+(?:\()+/;
export const attributeArgsRegex = /(?:\()+([A-Za-z])+\:+(.+)+(?:\))+/;

export function parseDocumentationAttributes<TData extends object = object>(
  documentation: string | undefined,
  expectedAttributeName: string,
  expectedAttributeKind: "model" | "field",
): Partial<TData> {
  const attributeRegex =
    expectedAttributeKind === "model"
      ? modelAttributeRegex
      : fieldAttributeRegex;
  for (const documentation_ of documentation?.split("\\n") ?? []) {
    const attribute = documentation_?.match(attributeRegex)?.[0];
    const attributeName = attribute
      ?.match(attributeNameRegex)?.[0]
      ?.slice(1, -1);
    if (attributeName !== expectedAttributeName) {
      continue;
    }
    const rawAttributeArgs = attribute
      ?.match(attributeArgsRegex)?.[0]
      ?.slice(1, -1);
    const splitRawArgsArray = rawAttributeArgs?.split(",").map(it => it.trim());
    const parsedAttributeArgs =
      splitRawArgsArray &&
      (Object.fromEntries(
        splitRawArgsArray.map(it => {
          const [key, value] = it.split(": ");
          return [key, JSON.parse(value)];
        }),
      ) as Partial<TData>);
    if (!parsedAttributeArgs) {
      continue;
    }
    return parsedAttributeArgs;
  }
  return {};
}
