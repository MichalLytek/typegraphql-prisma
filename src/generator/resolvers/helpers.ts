import { OptionalKind, MethodDeclarationStructure, Writers } from "ts-morph";

import { DmmfDocument } from "../dmmf/dmmf-document";
import { DMMF } from "../dmmf/types";
import { GeneratorOptions } from "../options";

export function generateCrudResolverClassMethodDeclaration(
  action: DMMF.Action,
  mapping: DMMF.ModelMapping,
  dmmfDocument: DmmfDocument,
  generatorOptions: GeneratorOptions,
): OptionalKind<MethodDeclarationStructure> {
  return {
    name: action.name,
    isAsync: true,
    returnType: `Promise<${action.returnTSType}>`,
    decorators: [
      {
        name: `TypeGraphQL.${action.operation}`,
        arguments: [
          `_returns => ${action.typeGraphQLType}`,
          Writers.object({
            nullable: `${!action.method.isRequired}`,
          }),
        ],
      },
    ],
    parameters: [
      {
        name: "ctx",
        // TODO: import custom `ContextType`
        type: "any",
        decorators: [{ name: "TypeGraphQL.Ctx", arguments: [] }],
      },
      {
        name: "info",
        type: "GraphQLResolveInfo",
        decorators: [{ name: "TypeGraphQL.Info", arguments: [] }],
      },
      ...(!action.argsTypeName
        ? []
        : [
            {
              name: "args",
              type: action.argsTypeName,
              decorators: [
                {
                  name: "TypeGraphQL.Args",
                  arguments: generatorOptions.emitRedundantDecoratorArgs
                    ? [`_returns => ${action.argsTypeName}`]
                    : [],
                },
              ],
            },
          ]),
    ],
    statements:
      action.kind === DMMF.ModelAction.aggregate
        ? [
            /* ts */ ` return getPrismaFromContext(ctx).${mapping.collectionName}.${action.prismaMethod}({
              ...args,
              ...transformInfoIntoPrismaArgs(info),
            });`,
          ]
        : action.kind === DMMF.ModelAction.groupBy
        ? [
            /* ts */ ` const { _count, _avg, _sum, _min, _max } = transformInfoIntoPrismaArgs(info);`,
            /* ts */ ` return getPrismaFromContext(ctx).${mapping.collectionName}.${action.prismaMethod}({
              ...args,
              ...Object.fromEntries(
                Object.entries({ _count, _avg, _sum, _min, _max }).filter(([_, v]) => v != null)
              ),
            });`,
          ]
        : [
            /* ts */ ` const { _count } = transformInfoIntoPrismaArgs(info);
            return getPrismaFromContext(ctx).${mapping.collectionName}.${action.prismaMethod}({
              ...args,
              ...(_count && transformCountFieldIntoSelectRelationsCount(_count)),
            });`,
          ],
  };
}
