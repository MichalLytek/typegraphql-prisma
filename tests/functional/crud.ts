import "reflect-metadata";
import { promises as fs } from "fs";
import { buildSchema } from "type-graphql";
import { graphql, GraphQLSchema } from "graphql";

import generateArtifactsDirPath from "../helpers/artifacts-dir";
import { generateCodeFromSchema } from "../helpers/generate-code";

describe("crud resolvers execution", () => {
  let outputDirPath: string;
  let graphQLSchema: GraphQLSchema;

  describe("basic operations", () => {
    beforeAll(async () => {
      outputDirPath = generateArtifactsDirPath("functional-crud");
      await fs.mkdir(outputDirPath, { recursive: true });
      const prismaSchema = /* prisma */ `
        model User {
          intIdField          Int     @id @default(autoincrement())
          uniqueStringField   String  @unique
          optionalStringField String?
          dateField           DateTime
        }
      `;
      await generateCodeFromSchema(prismaSchema, { outputDirPath });
      const { UserCrudResolver } = require(
        outputDirPath + "/resolvers/crud/User/UserCrudResolver.ts",
      );

      graphQLSchema = await buildSchema({
        resolvers: [UserCrudResolver],
        validate: false,
      });
    });

    it("should properly call PrismaClient on `findUnique` action", async () => {
      const document = /* graphql */ `
        query {
          user(where: { uniqueStringField: "uniqueValue" }) {
            intIdField
            dateField
          }
        }
      `;
      const prismaMock = {
        user: {
          findUnique: jest.fn().mockResolvedValue({
            intIdField: 1,
            dateField: new Date("2019-12-31T14:16:02.572Z"),
          }),
        },
      };

      const { data, errors } = await graphql({
        schema: graphQLSchema,
        source: document,
        contextValue: { prisma: prismaMock },
      });

      expect(errors).toBeUndefined();
      expect(data).toMatchSnapshot("user mocked response");
      expect(prismaMock.user.findUnique.mock.calls).toMatchSnapshot(
        "findUniqueUser call args",
      );
    });

    it("should properly call PrismaClient on `findMany` action", async () => {
      const document = /* graphql */ `
        query {
          users(
            take: 1
            skip: 1
            orderBy: { intIdField: desc }
            where: { dateField: { lte: "2019-12-31T19:16:02.572Z" } }
          ) {
            intIdField
            dateField
          }
        }
      `;
      const prismaMock = {
        user: {
          findMany: jest.fn().mockResolvedValue([
            {
              intIdField: 1,
              dateField: new Date("2019-12-31T14:16:02.572Z"),
            },
          ]),
        },
      };

      const { data, errors } = await graphql({
        schema: graphQLSchema,
        source: document,
        contextValue: { prisma: prismaMock },
      });

      expect(errors).toBeUndefined();
      expect(data).toMatchSnapshot("users mocked response");
      expect(prismaMock.user.findMany.mock.calls).toMatchSnapshot(
        "findManyUser call args",
      );
    });

    it("should properly call PrismaClient on `create` action", async () => {
      const document = /* graphql */ `
        mutation {
          createOneUser(
            data: {
              uniqueStringField: "unique"
              optionalStringField: "optional"
              dateField: "2019-12-31T14:16:02.572Z"
            }
          ) {
            intIdField
            dateField
          }
        }
      `;
      const prismaMock = {
        user: {
          create: jest.fn().mockResolvedValue({
            intIdField: 1,
            dateField: new Date("2019-12-31T14:16:02.572Z"),
          }),
        },
      };

      const { data, errors } = await graphql({
        schema: graphQLSchema,
        source: document,
        contextValue: { prisma: prismaMock },
      });

      expect(errors).toBeUndefined();
      expect(data).toMatchSnapshot("createUser mocked response");
      expect(prismaMock.user.create.mock.calls).toMatchSnapshot(
        "createUser call args",
      );
    });

    it("should properly call PrismaClient on `delete` action", async () => {
      const document = /* graphql */ `
        mutation {
          deleteOneUser(
            where: {
              uniqueStringField: "unique"
            }
          ) {
            intIdField
            dateField
          }
        }
      `;
      const prismaMock = {
        user: {
          delete: jest.fn().mockResolvedValue({
            intIdField: 1,
            dateField: new Date("2019-12-31T14:16:02.572Z"),
          }),
        },
      };

      const { data, errors } = await graphql({
        schema: graphQLSchema,
        source: document,
        contextValue: { prisma: prismaMock },
      });

      expect(errors).toBeUndefined();
      expect(data).toMatchSnapshot("deleteUser mocked response");
      expect(prismaMock.user.delete.mock.calls).toMatchSnapshot(
        "deleteUser call args",
      );
    });

    it("should properly call PrismaClient on `update` action", async () => {
      const document = /* graphql */ `
        mutation {
          updateOneUser(
            data: {
              dateField: {
                set: "2019-12-31T14:16:02.572Z",
              }
            }
            where: {
              uniqueStringField: "unique"
            }
          ) {
            intIdField
            dateField
          }
        }
      `;
      const prismaMock = {
        user: {
          update: jest.fn().mockResolvedValue({
            intIdField: 1,
            dateField: new Date("2019-12-31T14:16:02.572Z"),
          }),
        },
      };

      const { data, errors } = await graphql({
        schema: graphQLSchema,
        source: document,
        contextValue: { prisma: prismaMock },
      });

      expect(errors).toBeUndefined();
      expect(data).toMatchSnapshot("updateUser mocked response");
      expect(prismaMock.user.update.mock.calls).toMatchSnapshot(
        "updateUser call args",
      );
    });

    it("should properly call PrismaClient on `updateMany` action", async () => {
      const document = /* graphql */ `
        mutation {
          updateManyUser(
            data: {
              optionalStringField: null,
            }
            where: {
              dateField: { lte: "2019-12-31T19:16:02.572Z" }
            }
          ) {
            count
          }
        }
      `;
      const prismaMock = {
        user: {
          updateMany: jest.fn().mockResolvedValue({
            count: 3,
          }),
        },
      };

      const { data, errors } = await graphql({
        schema: graphQLSchema,
        source: document,
        contextValue: { prisma: prismaMock },
      });

      expect(errors).toBeUndefined();
      expect(data).toMatchSnapshot("updateManyUser mocked response");
      expect(prismaMock.user.updateMany.mock.calls).toMatchSnapshot(
        "updateManyUser call args",
      );
    });

    it("should properly call PrismaClient on `deleteMany` action", async () => {
      const document = /* graphql */ `
        mutation {
          deleteManyUser(
            where: {
              dateField: { lte: "2019-12-31T19:16:02.572Z" }
            }
          ) {
            count
          }
        }
      `;
      const prismaMock = {
        user: {
          deleteMany: jest.fn().mockResolvedValue({
            count: 3,
          }),
        },
      };

      const { data, errors } = await graphql({
        schema: graphQLSchema,
        source: document,
        contextValue: { prisma: prismaMock },
      });

      expect(errors).toBeUndefined();
      expect(data).toMatchSnapshot("deleteManyUser mocked response");
      expect(prismaMock.user.deleteMany.mock.calls).toMatchSnapshot(
        "deleteManyUser call args",
      );
    });

    it("should properly call PrismaClient on `upsert` action", async () => {
      const document = /* graphql */ `
        mutation {
          upsertOneUser(
            where: {
              uniqueStringField: "unique"
            }
            create: {
              uniqueStringField: "unique"
              optionalStringField: "optional"
              dateField: "2019-12-31T14:16:02.572Z"
            }
            update: {
              optionalStringField: null,
            }
          ) {
            intIdField
            uniqueStringField
            optionalStringField
            dateField
          }
        }
      `;
      const prismaMock = {
        user: {
          upsert: jest.fn().mockResolvedValue({
            intIdField: 1,
            uniqueStringField: "unique",
            optionalStringField: "optional",
            dateField: new Date("2019-12-31T14:16:02.572Z"),
          }),
        },
      };

      const { data, errors } = await graphql({
        schema: graphQLSchema,
        source: document,
        contextValue: { prisma: prismaMock },
      });

      expect(errors).toBeUndefined();
      expect(data).toMatchSnapshot("upsertUser mocked response");
      expect(prismaMock.user.upsert.mock.calls).toMatchSnapshot(
        "upsertUser call args",
      );
    });

    describe("_count for relations", () => {
      let outputDirPath: string;
      let graphQLSchema: GraphQLSchema;
      beforeAll(async () => {
        outputDirPath = generateArtifactsDirPath("functional-crud");
        await fs.mkdir(outputDirPath, { recursive: true });
        const prismaSchema = /* prisma */ `
          model FirstModel {
            idField            Int            @id @default(autoincrement())
            uniqueStringField  String         @unique
            floatField         Float
            secondModelsField  SecondModel[]
          }
          model SecondModel {
            idField            Int          @id @default(autoincrement())
            uniqueStringField  String       @unique
            floatField         Float
            firstModelFieldId  Int
            firstModelField    FirstModel   @relation(fields: [firstModelFieldId], references: [idField])
          }
        `;
        await generateCodeFromSchema(prismaSchema, { outputDirPath });
        const { FirstModelCrudResolver } = require(
          outputDirPath +
            "/resolvers/crud/FirstModel/FirstModelCrudResolver.ts",
        );

        graphQLSchema = await buildSchema({
          resolvers: [FirstModelCrudResolver],
          validate: false,
        });
      });

      it("should properly call prisma client", async () => {
        const document = /* graphql */ `
          query {
            findFirstFirstModel {
              idField
              _count {
                secondModelsField
              }
            }
          }
        `;
        const prismaMock = {
          firstModel: {
            findFirst: jest.fn().mockResolvedValue({
              idField: 1,
              _count: {
                secondModelsField: 2,
              },
            }),
          },
        };

        const { data, errors } = await graphql({
          schema: graphQLSchema,
          source: document,
          contextValue: { prisma: prismaMock },
        });

        expect(errors).toBeUndefined();
        expect(data).toMatchSnapshot("secondModelsField count mocked response");
        expect(prismaMock.firstModel.findFirst.mock.calls).toMatchSnapshot(
          "firstModel.findFirst call args",
        );
      });
    });
  });

  describe("aggregations", () => {
    beforeAll(async () => {
      outputDirPath = generateArtifactsDirPath("functional-crud");
      await fs.mkdir(outputDirPath, { recursive: true });
      const prismaSchema = /* prisma */ `
        model User {
          idField     Int  @id @default(autoincrement())
          intField    Int
          floatField  Int
        }
      `;
      await generateCodeFromSchema(prismaSchema, { outputDirPath });
      const { UserCrudResolver } = require(
        outputDirPath + "/resolvers/crud/User/UserCrudResolver.ts",
      );

      graphQLSchema = await buildSchema({
        resolvers: [UserCrudResolver],
        validate: false,
      });
    });

    it("should properly call PrismaClient on `aggregate` action with simple count field", async () => {
      const document = /* graphql */ `
        query {
          aggregateUser(
            take: 1
            skip: 1
            orderBy: { intField: desc }
            where: { floatField: { lte: 50 } }
          ) {
            _count {
              _all
            }
          }
        }
      `;
      const prismaMock = {
        user: {
          aggregate: jest.fn().mockResolvedValue({ _count: { _all: 5 } }),
        },
      };

      const { data, errors } = await graphql({
        schema: graphQLSchema,
        source: document,
        contextValue: { prisma: prismaMock },
      });

      expect(errors).toBeUndefined();
      expect(data).toMatchSnapshot("aggregateUserCount mocked response");
      expect(prismaMock.user.aggregate.mock.calls).toMatchSnapshot(
        "user.aggregate call args",
      );
    });

    it("should properly call PrismaClient on `aggregate` action with advanced operations", async () => {
      const document = /* graphql */ `
        query {
          aggregateUser(
            take: 1
            skip: 1
            orderBy: { intField: desc }
            where: { floatField: { lte: 50 } }
          ) {
            __typename
            _count {
              intField
              floatField
            }
            _min {
              __typename
              intField
              floatField
            }
            _max {
              __typename
              intField
              floatField
            }
            _sum {
              __typename
              intField
              floatField
            }
            _avg {
              __typename
              intField
              floatField
            }
          }
        }
      `;
      const prismaMock = {
        user: {
          aggregate: jest.fn().mockResolvedValue({
            _count: {
              intField: 1,
              floatField: 1,
            },
            _min: {
              intField: 0,
              floatField: 0,
            },
            _max: {
              intField: 10,
              floatField: 10,
            },
            _sum: {
              intField: 10,
              floatField: 10,
            },
            _avg: {
              intField: 5,
              floatField: 5,
            },
          }),
        },
      };

      const { data, errors } = await graphql({
        schema: graphQLSchema,
        source: document,
        contextValue: { prisma: prismaMock },
      });

      expect(errors).toBeUndefined();
      expect(data).toMatchSnapshot("aggregateUser mocked response");
      expect(prismaMock.user.aggregate.mock.calls).toMatchSnapshot(
        "user.aggregate call args",
      );
    });

    it("should properly call PrismaClient on `groupBy` action with advanced operations", async () => {
      const document = /* graphql */ `
        query {
          groupByUser(
            by: [intField]
            where: { floatField: { gte: 0 } }
          ) {
            __typename
            intField
            _count {
              __typename
              _all
            }
            _min {
              __typename
              intField
              floatField
            }
          }
        }
      `;
      const prismaMock = {
        user: {
          groupBy: jest.fn().mockResolvedValue([
            {
              intField: 10,
              _count: {
                _all: 5,
              },
              _min: {
                intField: 0,
                floatField: 0,
              },
            },
          ]),
        },
      };

      const { data, errors } = await graphql({
        schema: graphQLSchema,
        source: document,
        contextValue: { prisma: prismaMock },
      });

      expect(errors).toBeUndefined();
      expect(data).toMatchSnapshot("groupByUser mocked response");
      expect(prismaMock.user.groupBy.mock.calls).toMatchSnapshot(
        "user.groupBy call args",
      );
    });
  });
});
