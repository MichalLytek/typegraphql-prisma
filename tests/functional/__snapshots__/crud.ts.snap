// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`crud resolvers execution aggregations should properly call PrismaClient on \`aggregate\` action with advanced operations: aggregateUser mocked response 1`] = `
Object {
  "aggregateUser": Object {
    "__typename": "AggregateUser",
    "_avg": Object {
      "__typename": "UserAvgAggregate",
      "floatField": 5,
      "intField": 5,
    },
    "_count": Object {
      "floatField": 1,
      "intField": 1,
    },
    "_max": Object {
      "__typename": "UserMaxAggregate",
      "floatField": 10,
      "intField": 10,
    },
    "_min": Object {
      "__typename": "UserMinAggregate",
      "floatField": 0,
      "intField": 0,
    },
    "_sum": Object {
      "__typename": "UserSumAggregate",
      "floatField": 10,
      "intField": 10,
    },
  },
}
`;

exports[`crud resolvers execution aggregations should properly call PrismaClient on \`aggregate\` action with advanced operations: user.aggregate call args 1`] = `
Array [
  Array [
    Object {
      "_avg": Object {
        "floatField": true,
        "intField": true,
      },
      "_count": Object {
        "floatField": true,
        "intField": true,
      },
      "_max": Object {
        "floatField": true,
        "intField": true,
      },
      "_min": Object {
        "floatField": true,
        "intField": true,
      },
      "_sum": Object {
        "floatField": true,
        "intField": true,
      },
      "orderBy": Array [
        UserOrderByWithRelationInput {
          "intField": "desc",
        },
      ],
      "skip": 1,
      "take": 1,
      "where": UserWhereInput {
        "floatField": IntFilter {
          "lte": 50,
        },
      },
    },
  ],
]
`;

exports[`crud resolvers execution aggregations should properly call PrismaClient on \`aggregate\` action with simple count field: aggregateUserCount mocked response 1`] = `
Object {
  "aggregateUser": Object {
    "_count": Object {
      "_all": 5,
    },
  },
}
`;

exports[`crud resolvers execution aggregations should properly call PrismaClient on \`aggregate\` action with simple count field: user.aggregate call args 1`] = `
Array [
  Array [
    Object {
      "_count": Object {
        "_all": true,
      },
      "orderBy": Array [
        UserOrderByWithRelationInput {
          "intField": "desc",
        },
      ],
      "skip": 1,
      "take": 1,
      "where": UserWhereInput {
        "floatField": IntFilter {
          "lte": 50,
        },
      },
    },
  ],
]
`;

exports[`crud resolvers execution aggregations should properly call PrismaClient on \`groupBy\` action with advanced operations: groupByUser mocked response 1`] = `
Object {
  "groupByUser": Array [
    Object {
      "__typename": "UserGroupBy",
      "_count": Object {
        "__typename": "UserCountAggregate",
        "_all": 5,
      },
      "_min": Object {
        "__typename": "UserMinAggregate",
        "floatField": 0,
        "intField": 0,
      },
      "intField": 10,
    },
  ],
}
`;

exports[`crud resolvers execution aggregations should properly call PrismaClient on \`groupBy\` action with advanced operations: user.groupBy call args 1`] = `
Array [
  Array [
    Object {
      "_count": Object {
        "_all": true,
      },
      "_min": Object {
        "floatField": true,
        "intField": true,
      },
      "by": Array [
        "intField",
      ],
      "where": UserWhereInput {
        "floatField": IntFilter {
          "gte": 0,
        },
      },
    },
  ],
]
`;

exports[`crud resolvers execution basic operations _count for relations should properly call prisma client: firstModel.findFirst call args 1`] = `
Array [
  Array [
    Object {
      "include": Object {
        "_count": Object {
          "select": Object {
            "secondModelsField": true,
          },
        },
      },
    },
  ],
]
`;

exports[`crud resolvers execution basic operations _count for relations should properly call prisma client: secondModelsField count mocked response 1`] = `
Object {
  "findFirstFirstModel": Object {
    "_count": Object {
      "secondModelsField": 2,
    },
    "idField": 1,
  },
}
`;

exports[`crud resolvers execution basic operations should properly call PrismaClient on \`create\` action: createUser call args 1`] = `
Array [
  Array [
    Object {
      "data": UserCreateInput {
        "dateField": 2019-12-31T14:16:02.572Z,
        "optionalStringField": "optional",
        "uniqueStringField": "unique",
      },
    },
  ],
]
`;

exports[`crud resolvers execution basic operations should properly call PrismaClient on \`create\` action: createUser mocked response 1`] = `
Object {
  "createOneUser": Object {
    "dateField": "2019-12-31T14:16:02.572Z",
    "intIdField": 1,
  },
}
`;

exports[`crud resolvers execution basic operations should properly call PrismaClient on \`delete\` action: deleteUser call args 1`] = `
Array [
  Array [
    Object {
      "where": UserWhereUniqueInput {
        "uniqueStringField": "unique",
      },
    },
  ],
]
`;

exports[`crud resolvers execution basic operations should properly call PrismaClient on \`delete\` action: deleteUser mocked response 1`] = `
Object {
  "deleteOneUser": Object {
    "dateField": "2019-12-31T14:16:02.572Z",
    "intIdField": 1,
  },
}
`;

exports[`crud resolvers execution basic operations should properly call PrismaClient on \`deleteMany\` action: deleteManyUser call args 1`] = `
Array [
  Array [
    Object {
      "where": UserWhereInput {
        "dateField": DateTimeFilter {
          "lte": 2019-12-31T19:16:02.572Z,
        },
      },
    },
  ],
]
`;

exports[`crud resolvers execution basic operations should properly call PrismaClient on \`deleteMany\` action: deleteManyUser mocked response 1`] = `
Object {
  "deleteManyUser": Object {
    "count": 3,
  },
}
`;

exports[`crud resolvers execution basic operations should properly call PrismaClient on \`findMany\` action: findManyUser call args 1`] = `
Array [
  Array [
    Object {
      "orderBy": Array [
        UserOrderByWithRelationInput {
          "intIdField": "desc",
        },
      ],
      "skip": 1,
      "take": 1,
      "where": UserWhereInput {
        "dateField": DateTimeFilter {
          "lte": 2019-12-31T19:16:02.572Z,
        },
      },
    },
  ],
]
`;

exports[`crud resolvers execution basic operations should properly call PrismaClient on \`findMany\` action: users mocked response 1`] = `
Object {
  "users": Array [
    Object {
      "dateField": "2019-12-31T14:16:02.572Z",
      "intIdField": 1,
    },
  ],
}
`;

exports[`crud resolvers execution basic operations should properly call PrismaClient on \`findUnique\` action: findUniqueUser call args 1`] = `
Array [
  Array [
    Object {
      "where": UserWhereUniqueInput {
        "uniqueStringField": "uniqueValue",
      },
    },
  ],
]
`;

exports[`crud resolvers execution basic operations should properly call PrismaClient on \`findUnique\` action: user mocked response 1`] = `
Object {
  "user": Object {
    "dateField": "2019-12-31T14:16:02.572Z",
    "intIdField": 1,
  },
}
`;

exports[`crud resolvers execution basic operations should properly call PrismaClient on \`update\` action: updateUser call args 1`] = `
Array [
  Array [
    Object {
      "data": UserUpdateInput {
        "dateField": DateTimeFieldUpdateOperationsInput {
          "set": 2019-12-31T14:16:02.572Z,
        },
      },
      "where": UserWhereUniqueInput {
        "uniqueStringField": "unique",
      },
    },
  ],
]
`;

exports[`crud resolvers execution basic operations should properly call PrismaClient on \`update\` action: updateUser mocked response 1`] = `
Object {
  "updateOneUser": Object {
    "dateField": "2019-12-31T14:16:02.572Z",
    "intIdField": 1,
  },
}
`;

exports[`crud resolvers execution basic operations should properly call PrismaClient on \`updateMany\` action: updateManyUser call args 1`] = `
Array [
  Array [
    Object {
      "data": UserUpdateManyMutationInput {
        "optionalStringField": null,
      },
      "where": UserWhereInput {
        "dateField": DateTimeFilter {
          "lte": 2019-12-31T19:16:02.572Z,
        },
      },
    },
  ],
]
`;

exports[`crud resolvers execution basic operations should properly call PrismaClient on \`updateMany\` action: updateManyUser mocked response 1`] = `
Object {
  "updateManyUser": Object {
    "count": 3,
  },
}
`;

exports[`crud resolvers execution basic operations should properly call PrismaClient on \`upsert\` action: upsertUser call args 1`] = `
Array [
  Array [
    Object {
      "create": UserCreateInput {
        "dateField": 2019-12-31T14:16:02.572Z,
        "optionalStringField": "optional",
        "uniqueStringField": "unique",
      },
      "update": UserUpdateInput {
        "optionalStringField": null,
      },
      "where": UserWhereUniqueInput {
        "uniqueStringField": "unique",
      },
    },
  ],
]
`;

exports[`crud resolvers execution basic operations should properly call PrismaClient on \`upsert\` action: upsertUser mocked response 1`] = `
Object {
  "upsertOneUser": Object {
    "dateField": "2019-12-31T14:16:02.572Z",
    "intIdField": 1,
    "optionalStringField": "optional",
    "uniqueStringField": "unique",
  },
}
`;
